Cache start/end events in nccl.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122731
* __->__ #122732
* #122539
* #122538

Adding an event cache has two goals:
(1) lower the overhead of of issuing collectives
(2) removes cudaEventDestroy from the watchdog thread.
    If cuda gets stuck due to nccl, then cudaEventDestroy might
    hang. This has traditionally gotten the watchdog thread stuck,
    causing us to rely on a separate thread to make sure the watchdog
    thread makes progress. With this change, we probably do not need that
    thread anymore, but we can first check to see if we continue to find
    any stack traces suggesting a heartbeat timeout after we land this change.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

zdevito committed:
Cache start/end events in nccl.

Adding an event cache has two goals:
(1) lower the overhead of of issuing collectives
(2) removes cudaEventDestroy from the watchdog thread.
    If cuda gets stuck due to nccl, then cudaEventDestroy might
    hang. This has traditionally gotten the watchdog thread stuck,
    causing us to rely on a separate thread to make sure the watchdog
    thread makes progress. With this change, we probably do not need that
    thread anymore, but we can first check to see if we continue to find
    any stack traces suggesting a heartbeat timeout after we land this change.

[ghstack-poisoned]

wconstab reviewed (commented):
this LGTM. do you have some testing/instrumentation to confirm it actually works as intended?

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -381,6 +381,48 @@ std::future<bool> launchAsyncGilCheck() {
   return resultFuture;
 }
 
+class CUDAEventCache {
```
by convention, Can we move the class definition to .hpp file so that we can access the class anywhere in the cpp file ?

zdevito replied:
ProcessGroupNCCL.hpp is included in 5 compilation units, whereas code in ProcessGroupNCCL.cpp only is in one, so in cases where the code is only used locally I tend to avoid the header files to keep compile times down.

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -381,6 +381,48 @@ std::future<bool> launchAsyncGilCheck() {
+    at::cuda::CUDAEvent* event = nullptr;
+    {
+      std::lock_guard<std::mutex> lock(deviceCache.mutex);
+      auto& events = deviceCache.events[timing ? 1 : 0];
```
Will deviceCache.events[timing ? 1 : 0] return only 1 'event' based on timing? if so, why plurals?

zdevito replied:
deviceCache.events is two vector: a list of unused events without timing, and a list of unused events with timing. Since timing isn't strictly a global property (you can force timing on for a particular process group), we need to handle the situation where we fulfill both.

shuqiangzhang replied:
Oh, I got it now, maybe it is just to me, but this was the confusing part to me: std::vector<at::cuda::CUDAEvent*> events[2], it is using a combination of vector and array semantics. Maybe a vector of vector is better.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -381,6 +381,48 @@ std::future<bool> launchAsyncGilCheck() {
 
+class CUDAEventCache {
+ public:
+  CUDAEventCache() : caches_(at::cuda::device_count()) {}
```
ProcessGroupNCCL now supports single device per thread only. Would that help make the implementation here even simpler?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -442,10 +484,10 @@ ProcessGroupNCCL::WorkNCCL::WorkNCCL(
   }
-  ncclEndEvent_ = std::make_shared<at::cuda::CUDAEvent>(
-      enableTiming ? cudaEventDefault : cudaEventDisableTiming);
+  ncclEndEvent_ = CUDAEventCache::get().create(device.index(), enableTiming);
```
Any idea how big a performance difference there will be between disableTiming and enableTiming?

zdevito replied:
Here is one analysis I've seen:

https://github.com/harrism/cuda_event_benchmark

Using the items_per_second column, recording an event with timing is 2.5us but without timing it is .25us. Similarly it is about ~10x to cache the events vs cudaEventCreate/Destroy each time.

zdevito commented:
I tested this locally with prints to ensure I see events getting reused. But I am looking for ideas of how to actually test it better.

zdevito committed:
Update on "Cache start/end events in nccl."


Adding an event cache has two goals:
(1) lower the overhead of of issuing collectives
(2) removes cudaEventDestroy from the watchdog thread.
    If cuda gets stuck due to nccl, then cudaEventDestroy might
    hang. This has traditionally gotten the watchdog thread stuck,
    causing us to rely on a separate thread to make sure the watchdog
    thread makes progress. With this change, we probably do not need that
    thread anymore, but we can first check to see if we continue to find
    any stack traces suggesting a heartbeat timeout after we land this change.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

pritamdamania87 commented:
> (2) removes cudaEventDestroy from the watchdog thread.
If cuda gets stuck due to nccl, then cudaEventDestroy might
hang. This has traditionally gotten the watchdog thread stuck,
causing us to rely on a separate thread to make sure the watchdog
thread makes progress. With this change, we probably do not need that
thread anymore, but we can first check to see if we continue to find
any stack traces suggesting a heartbeat timeout after we land this change.

This isn't a reliable or general way to the solve the problem mentioned in https://github.com/pytorch/pytorch/issues/101463.  As mentioned in https://github.com/pytorch/pytorch/issues/101463, the watchdog thread also calls things like `isCompleted()`: https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1647 which eventually call things like `cudaGetLastError` which can get stuck too. There are probably a lot of CUDA calls that occur which might not be easy to find and track down in the watchdog thread. That is why a separate minimalistic thread can avoid such situations more reliably.

kwen2501 reviewed (approved):
LGTM!
